framework/include/materials/Material.h

```diff
@@ -258,6 +259,25 @@ class Material :
    */
   virtual QpData * createData();
 
+  /**
+   * Get a apecial MaterialUserObject which can be explicitly called by the material
+   * during the material property calculation.
+   * @param name The name of the parameter key of the user object to retrieve
+   * @return The user object with name associated with the parameter 'name'
+   */
+  template<class T>
+  const T & getMaterialUserObject(const std::string & name);
```

Comment:
I will add another opinion to these, getUserObject should just handle it. If a MaterialUserObject is requested, then use the tid, if it is not then don't. You are doing a dynamic_cast to MaterialUserObject in this function, so just do this in the regular getUserObject and have the cast determine what to do with the tid.